pkg/gadgets/dns/tracer/tracer.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"runtime"
+	"sync"
+	"syscall"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/perf"
+	log "github.com/sirupsen/logrus"
+	"github.com/vishvananda/netns"
+	"golang.org/x/sys/unix"
+)
+
+// #include "bpf/dns-common.h"
+import "C"
+
+const (
+	BPF_PROG_NAME = "bpf_prog1"
+	BPF_MAP_NAME  = "events"
+	SO_ATTACH_BPF = 50
+)
+
+type link struct {
+	collection *ebpf.Collection
+	perfRd     *perf.Reader
+
+	sockFd int
+}
+
+type Tracer struct {
+	mu sync.Mutex
+
+	spec *ebpf.CollectionSpec
+
+	// key: namespace/podname
+	// value: Tracelet
+	attachments map[string]*link
+}
+
+// Both openRawSock and htons are from github.com/cilium/ebpf:
+// MIT License
+// https://github.com/cilium/ebpf/blob/eaa1fe7482d837490c22d9d96a788f669b9e3843/example_sock_elf_test.go#L146-L166
+func openRawSock(pid uint32) (int, error) {
+	if pid != 0 {
+		// Lock the OS Thread so we don't accidentally switch namespaces
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
+
+		// Save the current network namespace
+		origns, _ := netns.Get()
+		defer origns.Close()
+
+		netnsHandle, err := netns.GetFromPid(int(pid))
+		if err != nil {
+			return -1, err
+		}
+		defer netnsHandle.Close()
+		err = netns.Set(netnsHandle)
+		if err != nil {
+			return -1, err
+		}
+
+		// Switch back to the original namespace
+		defer netns.Set(origns)
+	}
+
+	sock, err := syscall.Socket(syscall.AF_PACKET, syscall.SOCK_RAW|syscall.SOCK_NONBLOCK|syscall.SOCK_CLOEXEC, int(htons(syscall.ETH_P_ALL)))
+	if err != nil {
+		return -1, err
+	}
+	sll := syscall.SockaddrLinklayer{
+		Ifindex:  0, // 0 matches any interface
+		Protocol: htons(syscall.ETH_P_ALL),
+	}
+	if err := syscall.Bind(sock, &sll); err != nil {
+		return -1, err
+	}
+	return sock, nil
+}
+
+// htons converts the unsigned short integer hostshort from host byte order to network byte order.
```

Comment:
Does this comment need to be updated?

Reply:
I think it is still valid. What do you suggest?

Reply:
What is `hostshort`?

Reply:
I guess it was the name of the variable at some point. The same message is on
https://github.com/cilium/ebpf/blob/eaa1fe7482d837490c22d9d96a788f669b9e3843/example_sock_elf_test.go#L161

I will remove it.